crates/typst/src/eval/version.rs

```diff
@@ -0,0 +1,223 @@
+use crate::diag::{bail, StrResult};
+use crate::eval::Value;
+use ecow::EcoVec;
+use serde::{Serialize, Serializer};
+use std::cmp::Ordering;
+use std::fmt::{Debug, Display, Formatter};
+use std::hash::{Hash, Hasher};
+use std::iter::repeat;
+use std::str::FromStr;
+use typst::eval::Array;
+
+/// A version, with any number of components.
+///
+/// The list of components is semantically extended by an infinite list of zeros.
+/// This means that, for example, `0.8` is the same as `0.8.0`.
+/// As a special case, the empty version (that has no components at all) is the same as `0`, `0.0`, `0.0.0`, and so on.
+#[derive(Clone)]
+pub struct Version(EcoVec<VersionComponent>);
+
+// helper macro to make sure COMPONENT_NAMES and resolve_component_name never go out of sync
+macro_rules! component_names {
+    (@gen [$($arms:tt)*] [$($index:tt)*] [$($values:literal),*] $name:literal $(, $($rest:tt)*)?) => {
+        component_names!(@gen [$($arms)* $name => $($index)*,] [$($index)* + 1] [$($values,)* $name] $($($rest)*)?);
+    };
+    (@gen [$($arms:tt)*] [$($index:tt)*] [$($values:literal),*] /* end */) => {
+        // note: index is now the index after the end, i.e. the len
+        pub const COMPONENT_NAMES: [&'static str; { $($index)* }] = [$($values),*];
+        fn resolve_component_name(name: &str) -> Option<usize> {
+            Some(match name {
+                $($arms)*
+                _ => return None
+            })
+        }
+    };
+    ($($input:tt)*) => {
+        component_names!(@gen [] [0] [] $($input)*);
+    };
+}
+
+impl Version {
+    component_names!["major", "minor", "patch"];
+
+    pub fn new() -> Self {
+        Self(EcoVec::new())
+    }
+
+    fn get(&self, index: usize) -> Option<VersionComponent> {
+        self.0.get(index).copied()
+    }
+
+    /// Get a component of a version.
+    ///
+    /// Always non-negative. Returns `0` if the version isn't specified to the necessary length.
+    pub fn at(&self, index: i64) -> StrResult<i64> {
+        if index < 0 {
+            bail!("version component index out of bounds ({index})");
+        }
+        Ok(usize::try_from(index)
+            .ok()
+            .and_then(|i| self.get(i))
+            .unwrap_or_default()
+            .get())
+    }
+
+    /// Get a named component of a version.
+    ///
+    /// Always non-negative. Returns `0` if the version isn't specified to the necessary length.
+    pub fn component(&self, name: &str) -> StrResult<i64> {
+        match Self::resolve_component_name(name) {
+            Some(i) => Ok(self.get(i).unwrap_or_default().get()),
+            None => bail!("unknown version component"),
+        }
+    }
+
+    /// Get an iterator of the values of all named version components for a version.
+    pub fn named_components(&self) -> impl Iterator<Item = (&'static str, i64)> + '_ {
```

Comment:
what is this used for?

Reply:
Huh. Nothing, apparently. Must have been a left-over from development.